### PR TITLE
siteconfig: Add 'NewBuilder, Update, WaitForCondition...' and other methods

### PR DESCRIPTION
**Pull request by mpmaruthu**
Addding 'NewClusterInstanceBuilder, Update, WaitFor...Condition' methods in siteconfig package.

@klaskosk , @trewest , Please help to review this PR when you get a time. Thanks!

### CODE REVIEW

File: pkg/siteconfig/clusterinstance.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance.go:275]: @mpmaruthu It looks like the actual type of the condition you're waiting for changes in the name but not the code for the function. Could you just have one `WaitForCondition` function and then use conditions as variables?

File: pkg/siteconfig/clusterinstance.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance.go:207]:
```suggestion
			err := builder.Delete()
			builder.Definition.ResourceVersion = ""
```
There's an edge case that can cause this function to fail if you don't reset the resource version. Sometimes it's fine but I like to reset it anyway to prevent issues later

File: pkg/siteconfig/clusterinstance.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance.go:20]: I think I'd prefer the conditions being in the test suite. My reasoning is that conditions may change between releases (like cgu in 4.12)  so older ones are only available with older release branches of eco-goinfra. But cnf/ran always uses the latest version of eco-goinfra so we'd have to create our own variables for older conditions. If we're creating our own variables in the test suite then might as well only do that. Let me know if you disagree though I'm open to hearing alternatives

**Comment by kononovn** [comment]
@mpmaruthu Please re-base

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:20]: Good point, I do agree the proposed approach to declare "ClusterInstanceConditionType" as variables in test suite instead of declaring it in the API/package(siteconfig) code.

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:207]: Okay, updated. Thanks! The proposed change will be in place after submitting a new patch for review.

**Comment by kononovn** [comment]
@klaskosk Can you review?

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go]: Updated. Thanks!

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go]: Taken care. changes are updated accordingly.

File: pkg/siteconfig/clusterinstance_test.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance_test.go:19]:
```suggestion
	defaultClusterInstanceCondition = metav1.Condition{Type: "", Status: metav1.ConditionTrue}
```
and then could you please remove the variable from `clusterinstance.go`

Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance_test.go:19]: Okay, updated. Thanks! The proposed change will be in place after submitting a new patch for review.

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:224]: Okay, updated. Thanks! The proposed change will be in place after submitting a new patch for review.

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:151]: Okay, updated. Thanks! The proposed change will be in place after submitting a new patch for review.

File: pkg/siteconfig/clusterinstance.go
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:38]: Okay, updated. Thanks!

File: pkg/siteconfig/clusterinstance_test.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance_test.go:32]: You can just use `NewCIBuilder` directly in the test instead of creating the `generateCIBuilder` variable

File: pkg/siteconfig/clusterinstance_test.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance_test.go:76]: these should be `testCase.name` and `testCase.namespace`, respectively, which should fix the test failure

File: pkg/siteconfig/clusterinstance.go
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance.go:135]: apologies for not noticing earlier, this should be `return builder`, which should fix the panic in the unit tests

**Comment by klaskosk** [comment]
lgtm, once the linter error is fixed I'll approve

**Review by klaskosk — APPROVED** [review]
/lgtm

**Pull request merged**
